### ISSUE TITLE
Querying kubelet/apiserver metrics causes extremely high load on master

### ISSUE THREAD

**Issue by cryptk** [rancher/k3s#2108]
**Environmental Info:**
K3s Version: k3s version v1.18.6+k3s1 (6f56fa1d)

Node(s) CPU architecture, OS, and Version: 
Linux node-01 5.4.51-v8+ #1327 SMP PREEMPT Thu Jul 23 11:11:34 BST 2020 aarch64 GNU/Linux

Cluster Configuration:
4x Raspberry Pi 4B+ 8GB, single master, 3 workers, all nodes have 500GB SSD boot drives

**Describe the bug:**
When installing prometheus-operator I began having some extremely high CPU usage and load on the master node in my RPi4B+ cluster and occasional "pending" alerts for apiserver outages.  I was able to trace this down to queries for the kubelet/apiserver metrics taking a long time to return and generating some pretty high load.  The cluster doesn't even have any "real" workloads on it yet, and only 34 pods (openebs, prometheus-operator, nginx-ingress, metallb, nothing crazy) and the kubelet metrics are returning a little over 25K different metrics.

**Steps To Reproduce:**
run `kubectl get --raw /metrics` on a Raspberry pi 4B+

**Expected behavior:**
gathering metrics to assess the performance of a cluster should not excessively negatively impact the performance of a cluster

**Actual behavior:**
Almost 25,000 metrics are returned and the process takes almost 10 seconds (and sometimes more).  I have almost 13,000 metrics just with the name apiserver_request_duration_seconds_bucket

**Comment by brandond** [comment]
@cryptk the apiserver metrics are all just native kubernetes functionality, this is not strictly part of k3s or something that we would modify.

The request duration metrics do have high cardinality because there are separate stats for each resource type, operation, and bucket. If you don't want to scrape these, you should be able to modify your prometheus config to skip them.

**Comment by cryptk** [comment]
So the issue lies upstream... ugh.  Is it normal/expected for scraping this endpoint to generate this kind of load on a RPi?

**Comment by cryptk** [comment]
I definitely believe you that this isn't really a k3s issue, but one thing I just found that is pretty interesting... I added a taint to the master node and everything got MUCH better... Then I removed the taint and allowed a single pod (an openebs pvc replica) to start on the master and the metrics endpoint started taking multiple seconds.

**Comment by brandond** [comment]
Have you compared cpu and io utilization on the server node with and without the OpenEBS pod on the node? Have you observed the same behavior with other pods running on the master?

Just as a side note, I found that OpenEBS has some of the worst io throughput figures of all the clustered storage options. Putting Prometheus on OpenEBS just destroyed io latency for everything else on the same backing disks. I'm still on it for my lab, but plan to evaluate a couple other options soon.

**Comment by cryptk** [comment]
@brandond I was able to replicate the issue by just running several nginx containters on the master.  There were no other containers on the master at the time as the master was tainted and the sts for the nginx pods had a toleration set.

It did require more containers, but the nginx containers were not doing anything, they were sitting completely idle.  There were 20 nginx containers on the master, all sitting completely idle, and the kubelet /metrics endpoint was taking about 10 seconds to load.  With no containers on the master the kubelet /metrics endpoint loads in a few hundred ms (typically 0.2 - 0.3 seconds).

I think the problem might be made worse if the pods on the master have things like secrets/configmaps/etc in use, but that will require further testing... that would explain though why it took 20 "barebones" nginx containers, but a more normal load of containers took far fewer pods.  I will also test again with a "normal" load of containers, just without any OpenEBS pvc pods.

I am also not super impressed with OpenEBS so far, I would love to run something like Longhorn, but there aren't arm64 builds for it out yet.

**Comment by cryptk** [comment]
So, I did a bunch more experimenting and I'm no closer to knowing the direct cause of this high load, but I have figured out what I can do to fix it.  I redid my k3s cluster using docker instead of containerd and I am now able to run pods on the k3s master node without the metrics endpoint dragging the whole system down.

With the same load as before (well, similar, maybe slightly fewer pods on the leader) I am still getting a large number of metrics (about 25k) but the endpoint is able to consistently return those metrics in about 0.5s (far better than the 10-15 seconds from before).

I suspect this has something to do with docker running with overlayfs as opposed to however containerd stores it's files and that enables faster disk usage/inode checks, but that is conjecture and I don't have any solid evidence pointing to that.

When running with docker instead of containerd I am finally able to schedule pods on the leader AND properly monitor the cluster.
